libdrgn/program.c

```diff
@@ -1108,6 +1108,140 @@ drgn_program_find_symbol_by_address(struct drgn_program *prog, uint64_t address,
+			return true;
+	if (arg->flags & SEARCH_ADDRESS)
+		if (arg->address >= addr && arg->address < addr + sym->st_size)
+			return true;
+	return false;
+}
+
+static int symbols_search_cb(Dwfl_Module *dwfl_module, void **userdatap,
+			     const char *module_name, Dwarf_Addr base,
+			     void *cb_arg)
+{
+	struct symbols_search_arg *arg = cb_arg;
+	int symtab_len, i;
+
+	symtab_len = dwfl_module_getsymtab(dwfl_module);
+	if (symtab_len == -1) {
+		arg->bad_symtabs = true;
+		return DWARF_CB_OK;
+	}
+	/* Ignore the zeroth null symbol */
+	for (i = 1; i < symtab_len; i++) {
+		GElf_Sym elf_sym;
+		GElf_Addr elf_addr;
+		const char *name;
+		struct drgn_symbol *sym;
+
+		name = dwfl_module_getsym_info(dwfl_module, i, &elf_sym,
+					       &elf_addr, NULL, NULL, NULL);
+		if (!symbol_match(arg, elf_addr, &elf_sym, name))
+			continue;
+
+		sym = malloc(sizeof(*sym));
+		if (!sym) {
+			arg->err = &drgn_enomem;
+			return DWARF_CB_ABORT;
+		}
+		drgn_symbol_from_elf(name, elf_addr, &elf_sym, sym);
+		if (!symbolp_vector_append(&arg->results, &sym)) {
+			drgn_symbol_destroy(sym);
+			arg->err = &drgn_enomem;
+			return DWARF_CB_ABORT;
+		}
+	}
+	return DWARF_CB_OK;
+}
+
+LIBDRGN_PUBLIC struct drgn_error *
+drgn_program_find_symbols_by_name(struct drgn_program *prog, const char *name,
+				  struct drgn_symbol ***syms_ret,
+				  size_t *count_ret)
+{
+	struct symbols_search_arg arg;
+
+	if (!prog->dbinfo)
+		return drgn_error_create(DRGN_ERROR_MISSING_DEBUG_INFO,
+					 "could not find matching symbols");
+
+	arg.name = name;
+	if (name)
+		arg.flags = SEARCH_NAME;
+	else
+		arg.flags = SEARCH_ALL;
+	symbolp_vector_init(&arg.results);
+
+	if (dwfl_getmodules(prog->dbinfo->dwfl, symbols_search_cb,
+			    &arg, 0)) {
+		for (size_t i = 0; i < arg.results.size; i++)
+			drgn_symbol_destroy(arg.results.data[i]);
+		symbolp_vector_deinit(&arg.results);
+		return arg.err;
+	}
+
+	symbolp_vector_shrink_to_fit(&arg.results);
+	*count_ret = arg.results.size;
+	*syms_ret = arg.results.data;
+	return NULL;
+}
+
+LIBDRGN_PUBLIC struct drgn_error *
+drgn_program_find_symbols_by_address(struct drgn_program *prog,
+				     uint64_t address,
+				     struct drgn_symbol ***syms_ret,
+				     size_t *count_ret)
+{
+	Dwfl_Module *module;
+	struct symbols_search_arg arg = {
+		.address = address,
+		.flags = SEARCH_ADDRESS,
+	};
+
+	if (!prog->dbinfo)
+		return drgn_error_create(DRGN_ERROR_MISSING_DEBUG_INFO,
+					 "could not find matching symbols");
+
+	symbolp_vector_init(&arg.results);
+	module = dwfl_addrmodule(prog->dbinfo->dwfl, address);
+	if (module) {
+		if (symbols_search_cb(module, NULL, NULL, 0, &arg) != DWARF_CB_OK) {
+			for (size_t i = 0; i < arg.results.size; i++)
+				drgn_symbol_destroy(arg.results.data[i]);
+			symbolp_vector_deinit(&arg.results);
+			return arg.err;
+		}
+	}
+	symbolp_vector_shrink_to_fit(&arg.results);
+	*count_ret = arg.results.size;
+	*syms_ret = arg.results.data;
+	return NULL;
```

Comment:
Other than how `arg` is initialized, this function is identical to `drgn_program_find_symbols_by_name()`, right? Can you refactor them to a common helper?

Reply:
They're not quite the same -- `drgn_program_find_symbols_by_name()` uses `dwfl_getmodules()` to iterate over each module and call `symbols_search_cb()` on them. Whereas this one uses `dwfl_addrmodule()` to identify the module containing the given address, and then it directly calls `symbols_search_cb()` on just that module.

I did previously have these two sharing a common helper (`symbols_search()`), but that resulted in searching every module for the case where `SYMBOL_SEARCH_ADDR` was set, and we decided to change that.

But I think you have a point here, I can re-introduce `symbols_search()` as a helper, but it can just choose whether to use `dwfl_addrmodule()` or `dwfl_getmodules()` depending on the value of `args`. That reuses most of the code, and it would allow for more general search interfaces if it was ever necessary.

Reply:
Duh, sorry about that, I missed that major difference which was exactly what we had just discussed. Yes, the `symbols_search()` helper you just described makes sense, so we can at least share the error handling and shrink/return code.